evap/staff/importers/enrollment.py

```diff
@@ -0,0 +1,764 @@
+        self.tracker: FirstLocationAndCountTracker = FirstLocationAndCountTracker()
+
+    def check_course_data(self, course_data: CourseData, location: ExcelFileLocation):
+        if not all_fields_valid(course_data):
+            return
+
+        stored = self.course_data_by_name_en.setdefault(course_data.name_en, course_data)
+
+        # degrees will be merged together
+        if course_data.equals_when_ignoring_degrees(stored):
+            return
+
+        self.tracker.add_location_for_key(location, course_data.name_en)
+
+    def finalize(self) -> None:
+        for key, location_string in self.tracker.aggregated_keys_and_location_strings():
+            self.importer_log.add_error(
+                _('{location}: The data of course "{name}" differs from its data in a previous row.').format(
+                    location=location_string,
+                    name=key,
+                ),
+                category=ImporterLogEntry.Category.COURSE,
+            )
+
+
+class TooManyEnrollmentsChecker(Checker, RowCheckerMixin):
+    """Warn when users exceed settings.IMPORTER_MAX_ENROLLMENTS enrollments"""
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.evaluation_names_en_per_user: Dict[str, Set[str]] = defaultdict(set)
+
+    def check_row(self, row: EnrollmentParsedRow):
+        if row.student_data.email == "":
+            return
+
+        self.evaluation_names_en_per_user[row.student_data.email].add(row.course_data.name_en)
+
+    def finalize(self) -> None:
+        for email, evaluation_names_en in self.evaluation_names_en_per_user.items():
+            enrollment_count = len(evaluation_names_en)
+            if enrollment_count > settings.IMPORTER_MAX_ENROLLMENTS:
+                self.importer_log.add_warning(
+                    _("Warning: User {} has {} enrollments, which is a lot.").format(email, enrollment_count),
+                    category=ImporterLogEntry.Category.MANY,
+                )
+
+
+class UserDataAdapter(RowCheckerMixin):
+    def __init__(self, user_data_checker) -> None:
+        self.user_data_checker = user_data_checker
+
+    def check_row(self, row: EnrollmentParsedRow) -> None:
+        self.user_data_checker.check_userdata(row.student_data, row.location)
+        self.user_data_checker.check_userdata(row.responsible_data, row.location)
+
+    def finalize(self) -> None:
+        self.user_data_checker.finalize()
+
+
+class CourseDataAdapter(RowCheckerMixin):
+    def __init__(self, course_data_checker) -> None:
+        self.course_data_checker = course_data_checker
+
+    def check_row(self, row: EnrollmentParsedRow) -> None:
+        self.course_data_checker.check_course_data(row.course_data, row.location)
+
+    def finalize(self) -> None:
+        self.course_data_checker.finalize()
+
+
+@transaction.atomic
+def import_enrollments(
+    excel_content, semester: Semester, vote_start_datetime, vote_end_date, test_run: bool
```

Comment:
what is holding us back from the following?
```python
def import_enrollments(
    excel_content: bytes,
    semester: Semester,
    vote_start_datetime: Optional[datetime],
    vote_end_date: Optional[date],
    test_run: bool,
) -> ImporterLog:
```

accordingly, `update_existing_and_create_new_courses` should be annotated with non-optional versions